### PR TITLE
ARROW-9702: [C++] Register bpacking SIMD to runtime path.

### PR DESCRIPTION
**Pull request by frankdjx**
1. Add AVX2 version for unpack32.
2. Add dynamic dispatch facility.
3. Pick the SIMD path at runtime by the CPU feature.

Signed-off-by: Frank Du <frank.du@intel.com>

### CODE REVIEW

**Comment by jianxind** [comment]
@ursabot benchmark --suite-filter=parquet-encoding-benchmark --benchmark-filter=BM_PlainDecodingBoolean

**Comment by jianxind** [comment]
> @ursabot benchmark --suite-filter=parquet-encoding-benchmark --benchmark-filter=BM_PlainDecodingBoolean

The benchmark machine not working now. Attach the data on my setup
Before:
```
BM_PlainDecodingBoolean/1024         618 ns          618 ns      1127308 bytes_per_second=1.5435G/s
BM_PlainDecodingBoolean/4096        2077 ns         2075 ns       337435 bytes_per_second=1.83799G/s
BM_PlainDecodingBoolean/32768      15613 ns        15597 ns        45041 bytes_per_second=1.95666G/s
BM_PlainDecodingBoolean/65536      31036 ns        31004 ns        22619 bytes_per_second=1.96861G/s
```
AVX2:
```
BM_PlainDecodingBoolean/1024         314 ns          314 ns      2232819 bytes_per_second=3.0417G/s
BM_PlainDecodingBoolean/4096         925 ns          925 ns       755970 bytes_per_second=4.12579G/s
BM_PlainDecodingBoolean/32768       6602 ns         6595 ns       106197 bytes_per_second=4.62716G/s
BM_PlainDecodingBoolean/65536      13085 ns        13072 ns        53606 bytes_per_second=4.669G/s
```
AVX512:
```
BM_PlainDecodingBoolean/1024         284 ns          283 ns      2470664 bytes_per_second=3.36689G/s
BM_PlainDecodingBoolean/4096         794 ns          794 ns       880595 bytes_per_second=4.80645G/s
BM_PlainDecodingBoolean/32768       5406 ns         5401 ns       129007 bytes_per_second=5.65012G/s
BM_PlainDecodingBoolean/65536      10688 ns        10677 ns        65521 bytes_per_second=5.71639G/s
```

**Comment by kszucs** [comment]
@ursabot benchmark --suite-filter=parquet-encoding-benchmark --benchmark-filter=BM_PlainDecodingBoolean

**Comment by pitrou** [comment]
Here are some benchmarks on an AMD Ryzen CPU (AVX2-enabled):
* before:
```
BM_PlainDecodingBoolean/1024         344 ns          344 ns      2030607 bytes_per_second=2.77389G/s
BM_PlainDecodingBoolean/4096        1084 ns         1084 ns       640969 bytes_per_second=3.52035G/s
BM_PlainDecodingBoolean/32768       7896 ns         7895 ns        88481 bytes_per_second=3.86522G/s
BM_PlainDecodingBoolean/65536      15708 ns        15706 ns        44567 bytes_per_second=3.88619G/s
```

**Review by pitrou — COMMENTED** [review]
Thank you for doing this. I'm not looking into the AVX2 implementation, but here are some comments.

**Comment by jianxind** [comment]
> @ursabot benchmark --suite-filter=parquet-encoding-benchmark --benchmark-filter=BM_PlainDecodingBoolean

Below is the results on this action https://ci.ursalabs.org/#/builders/73/builds/100.

```
                       benchmark       baseline      contender  change %                                           counters
2  BM_PlainDecodingBoolean/65536  1.277 GiB/sec  3.120 GiB/sec   144.247  {'run_name': 'BM_PlainDecodingBoolean/65536', ...
0  BM_PlainDecodingBoolean/32768  1.273 GiB/sec  3.097 GiB/sec   143.296  {'run_name': 'BM_PlainDecodingBoolean/32768', ...
1   BM_PlainDecodingBoolean/4096  1.222 GiB/sec  2.438 GiB/sec    99.514  {'run_name': 'BM_PlainDecodingBoolean/4096', '...
3   BM_PlainDecodingBoolean/1024  1.049 GiB/sec  1.838 GiB/sec    75.251  {'run_name': 'BM_PlainDecodingBoolean/1024', '...
```

File: cpp/src/arrow/util/bpacking.cc [hunk 1]
```diff
@@ -0,0 +1,174 @@
...
+      DCHECK(false) << "Unsupported num_bits";
+  }
+
+  return batch_size;
+}
+
+typedef int (*unpack32_func_t)(const uint32_t* in, uint32_t* out, int batch_size,
+                               int num_bits);
+
+static std::once_flag unpack32_ptr_initialized;
```
[the remark this answers is not in the document]
Comment (jianxind) [reply cpp/src/arrow/util/bpacking.cc:149]: Thanks, I create a JIRA, https://issues.apache.org/jira/browse/ARROW-9717. Will work on the facility later.

[the remark this answers is not in the document]
Comment (jianxind) [reply cpp/src/arrow/util/bpacking.cc]: Hi Antoine, just draft a dispatch facility, can you help to review it again?

**Comment by pitrou** [comment]
@jianxind I've made some slight changes to the dynamic dispatch facility (also, I rebased and force-pushed). Can you tell me if the changes are ok to you?

**Review by pitrou — APPROVED** [review]
+1

**Comment by jianxind** [comment]
> @jianxind I've made some slight changes to the dynamic dispatch facility (also, I rebased and force-pushed). Can you tell me if the changes are ok to you?

@pitrou  Thanks for the optimization, the dispatch code is clear now, I really learn a lot here, thank you. I verified it works well on my setup.

**Pull request closed without merging**